substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/isolated/IsolatedSpeculationLog.java

```diff
@@ -48,19 +54,63 @@ public void collectFailedSpeculations() {
         collectFailedSpeculations0(IsolatedCompileContext.get().getClient(), handle);
     }
 
+    class SpeculationReasonInvocationHandler implements InvocationHandler {
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            return IsolatedSpeculationReasonEncoding.class.getDeclaredMethod(method.getName(), method.getParameterTypes()).invoke(args);
+        }
+    }
+
+    class SpeculationReasonEncodingSupplier implements Supplier<Object> {
+        private Object speculationReasonEncodingProxy;
+
+        SpeculationReasonEncodingSupplier(Object speculationReasonEncodingProxy) {
+            this.speculationReasonEncodingProxy = speculationReasonEncodingProxy;
+        }
+
+        @Override
+        public Object get() {
+            try {
+                return speculationReasonEncodingProxy.getClass().getDeclaredConstructor().newInstance();
+            } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+                e.printStackTrace();
+            }
+            return null;
+        }
+    }
+
+    private byte[] getByteArray(SpeculationReason reason) {
+        byte[] bytes = null;
+        if (reason instanceof UnencodedSpeculationReason) {
+            Object encoding = ((UnencodedSpeculationReason) reason).encode();
+            assert encoding instanceof IsolatedSpeculationReasonEncoding;
+            bytes = ((IsolatedSpeculationReasonEncoding) encoding).getByteArray();
+        } else {
+            try {
+                Class<?> speculationReasonEncodingClass = Class.forName("jdk.vm.ci.meta.SpeculationLog$SpeculationReasonEncoding");
+                final Object speculationReasonEncodingProxy = Proxy.newProxyInstance(speculationReasonEncodingClass.getClassLoader(),
```

Comment:
Proxy and reflection accessors will eventually be cached in static variables as requested in https://github.com/oracle/graal/issues/2933#issuecomment-714362574

Reply:
Please do cache the proxy class `Proxy.newProxyInstance(...).getClass()` in a `static final` field. Reflection accessors do not need to be cached as long as its arguments are compile-time constants.